### DISCUSSION TITLE
Disabling logger in a test context

### DISCUSSION THREAD

**Discussion by alanjaouen** [jongpie/NebulaLogger#484] [Q&A] [answered]
Hi,

I'm wondering how to disable the Logger behaviour during UnitTesting, to reduce CPU time usage,

eg this is one overview of a unit test
![image](https://user-images.githubusercontent.com/14994179/233995526-3c390cc0-ee19-4cee-808c-9ccb9200ef74.png)
I tried to use `Logger.suspendSaving();`
and it save some DB usage
![image](https://user-images.githubusercontent.com/14994179/233995800-3e92dcb4-a0e9-460e-ac4d-99bb4dc024d3.png)

but I still have some minor DB usage with  `14:00:11:079 DML_BEGIN [304]|Op:Insert|Type:LogEntryEvent__e|Rows:4`

Is there a way to disable insertion of LogEntryEvent__e during Unit tests?

Best Regards,

**Comment by jongpie** [comment]
Hi @alanjaouen - any chance you could provide a sample of what your test is doing / how you're calling `Logger.suspendSaving()` in your test? And are you inserting/updating any `LoggerSettings__c` records in your test? In theory, using `Logger.suspendSaving()` should prevent any DB usage, but there could be a bug somewhere in Nebula Logger's code.

Another way to approach it is to create a `LoggerSettings__c` record in your test (or in a `@TestSetup` method) with `IsEnabled__c = false`. Something like this would work for a single test method:

```java
@IsTest
private static void runsSomeTest() {
  LoggerSettings__c settings = LoggerSettings__c.getOrgDefaults();
  settings.IsEnabled__c = false;
  insert settings;
  
  // TODO add your test here - logging will now be disabled in this test method
}
```

**Comment by alanjaouen** [comment]
Hi @jongpie 

here is a sample of what i'm doing
```apex
@isTest(isParallel=true)
public class myTestClass {
    @testSetup
    public static void setup() {
        // disable log saving
        Logger.suspendSaving();
    }

    @isTest
    public static void convertInvoiceToJsonWithOneId() {
        // Given - data
        Id id = fflib_IDGenerator.generate(blng__Invoice__c.SObjectType);
        Set<Id> invoiceIds = new Set<Id>{ id };

        // When
        String returned = new myService().MyMethod(invoiceIds);

        // Then
        // some asserts
    }

    ...
}
```

For what I see, we insert events `LogEntryEvent__e`, and logs reccords are not saved to DB.
I think that to remove the DB usage that I see I should prevent the insertion of these events?

**Reply by jongpie** [reply, marked as answer]
Hi @alanjaouen - using `Logger.suspendSaving()` within `@TestSetup` won't work unfortunately, due to how Apex resets some transaction limits & variables before running your `@IsTest` method. You could instead use this in your `@TestSetup` method to disable logging in all of your tests:

```java
@isTest(isParallel=true)
public class myTestClass {
    @testSetup
    public static void setup() {
        // disable log saving
        LoggerSettings__c settings = LoggerSettings__c.getOrgDefaults();
        settings.IsEnabled__c = false;
        upsert settings;
    }

    @isTest
    public static void convertInvoiceToJsonWithOneId() {
        // Given - data
        Id id = fflib_IDGenerator.generate(blng__Invoice__c.SObjectType);
        Set<Id> invoiceIds = new Set<Id>{ id };

        // When
        String returned = new myService().MyMethod(invoiceIds);

        // Then
        // some asserts
    }

    ...
}
```

Let me know if this works for you!
